packages/babel-parser/src/tokenizer/index.js

```diff
@@ -123,13 +127,25 @@ export default class Tokenizer extends LocationParser {
     this.isLookahead = false;
   }
 
+  pushToken(token: Token | N.Comment) {
+    // Pop out invalid tokens trapped by try-catch parsing.
+    // Those parsing branches are mainly created by typescript and flow plugins.
+    while (
+      this.tokens.length > 0 &&
+      this.tokens[this.tokens.length - 1].end > token.start
+    ) {
+      this.tokens.pop();
+    }
```

Comment:
To make this faster, we can store `this.state.tokensLength: number`.
Then, when we `pushToken`, we first set `this.tokens.length = this.state.tokensLength`.
By doing so we can delete the wrong tokens in `O(1)` time rather than looping over the list.

Reply:
I thought about that.

If some future plugin calls `this.state = oldState` directly without knowing to restore tokensLength, that will cause trouble.

Now I think I can hide the details in `get state()` and `set state(state)`, that requires no code changes in plugins.

Reply:
`state` is accessed a lot, I don't know how much a getter will impact performance.

> If some future plugin calls this.state = oldState directly without knowing to restore tokensLength, that will cause trouble.

`oldState` will already have the old `tokensLength`, so it would be automatically restored.

Reply:
I just missed your example. Is it better to restore the length when set state, or just before push token?

Reply:
I think that it's ok to do it in `pushToken` since it will always be called.
My suggestion was to avoid looping over the existing tokens list and checking their `start`/`end` position.

Reply:
I have made the suggested change.